brave/brave-core
# Prototype pollution prevention for wallet scripts

Resolves https://github.com/brave/brave-browser/issues/24415
Resolves https://github.com/brave/security/issues/940

1. Provide `extensions::SafeBuiltins` like class for Brave wallet and only save what we need. (Most of the codes are copied from upstream)
2. Move unnecessary Object creation from script to v8
3. This measure is to protect the scripts we loaded in Brave, 3rd party scripts are out of scope
4. SafeBuiltins are loaded in a closure when executing the script so it won't pollute the global namespace  

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
1. Setup wallet and solana account
5. Navigate to simple website like [example.com](https://example.com/)
6. Open Dev console
7. Type this and enter
```
Object.defineProperties = ()=>console.log('defineProperties overwritten')
await solana.connect()
_brave_solana.createPublickey = ()=>'0x12345'
_brave_solana.createPublickey('5STiJLNtCLiNaeD6vwrEfvikNi9NbfVpEbxU7KrUv4uo').toString()
```
8. Grant permission or unlock the wallet when being asked.
9. We should expect to see `'5STiJLNtCLiNaeD6vwrEfvikNi9NbfVpEbxU7KrUv4uo'`
10. Refresh the page
11. Type this and enter 
```
await solana.connect()
Object.prototype.__defineGetter__('publicKey', ()=>'0x00')
solana.publicKey
```
12. We should be able to see real public key instead of '0x00'
13.  Refresh the page
14. Type this and enter 
```
Object.freeze = ()=>{}
await solana.connect()
_brave_solana.abc = 123
_brave_solana.abc
```
15. We should see undefined instead of 123
